pkg/puller/downloader.go

```diff
@@ -0,0 +1,173 @@
+package puller
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"hash/fnv"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+var (
+	d *Downloader
+)
+
+type Downloader struct {
+	NumRetries int
+	ModelDir   string
+}
+
+type Protocol string
+
+const (
+	// Supported Protocols
+	S3    Protocol = "s3://"
+	GS    Protocol = "gs://"
+	PVC   Protocol = "pvc://"
```

Comment:
I would suggest to only keep s3 and gs protocol for now and add other protocols when we do support them.